flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -210,6 +221,21 @@ public void registerSharedStatesAfterRestored(SharedStateRegistry sharedStateReg
 	//  Discard and Dispose
 	// ------------------------------------------------------------------------
 
+	/**
+	 * Asynchronously call a discard on the ioExecutor
+	 * (FixedThreadPool of configurable size of default 4*CPU cores)
+	 * and count the number of checkpoints that are waiting to clean.
+	 */
+	void asyncDiscardCheckpointAndCountCheckpoint(ThrowingConsumer<CompletedCheckpoint, Exception> discardCallback, Executor executor){
+		cleanCallback.accept(() -> {
+			try {
+				discardCallback.accept(this);
```

Comment:
This callback calling callback and `callback.accept(this)` doesn't look very readable to me. 

Looking at `discardCallback`, I see:
1. it's a call to `CompletedCheckpoint.doDiscard` if some condition is `true`
1. the condition can be evaluated early

Therefore, we can get rid of it by:
1. evaluating condition ASAP in `ZooKeeperCompletedCheckpointStore` (maybe extract methods `CompletedCheckpoint.shouldDiscardOnShutdown` and `OnSubsume`)
1. passing `boolean discard` to `tryRemoveCompletedCheckpoint` and further instead of callback
1. replacing `discardCallback.accept(this);`
with `if (discard) this.doDiscard();`. 
in `CompletedCheckpoint.asyncDiscardCheckpointAndCountCheckpoint`:

WDYT?

Reply:
This is the code you gave at the end of [your comment](https://github.com/apache/flink/pull/13040#issuecomment-686639977) I just reused it. But anyway no problem about trying to improve it.

Regarding your proposed refactoring: I understand that in place of passing callbacks `CompletedCheckpoint::discardOnSubsume` and `CompletedCheckpoint::discardOnShutdown`, as they both call `CompletedCheckpoint.doDiscard()` you prefer calling it directly. But it would require to evaluate the conditions inside ZooKeeperStore:

- for `CompletedCheckpoint::discardOnSubsume`: means moving condition `if (props.discardOnSubsumed()) `

- for `CompletedCheckpoint::discardOnShutdown`: means moving conditions `if (jobStatus == JobStatus.FINISHED && props.discardOnJobFinished() ||
				jobStatus == JobStatus.CANCELED && props.discardOnJobCancelled() ||
				jobStatus == JobStatus.FAILED && props.discardOnJobFailed() ||
				jobStatus == JobStatus.SUSPENDED && props.discardOnJobSuspended()) `
=> I'm not sure it's worth it, violating the integrity of CompletedCheckpoint object for the sake a readable `asyncDiscardCheckpointAndCountCheckpoint` method.

Reply:
Sure, we shouldn't break the encapsulation of `CompletedCheckpoint`, but we could extract these conditions into public methods (`boolean shouldDiscardOnXxx()`) which would be used by ZooKeeperStore.
If you think it doesn't worth it I'm also fine with the current approach.

Reply:
I tried this refactoring and I find weird to evaluate the conditions at several places I find it more clear to have conditions evaluated in the `discardOn*` methods and have these methods called as a whole. I prefer this even if it means having unnatural `discardCallback.accept(this);` So if you don't mind I'll leave it as it is.